Track metadata improvements and fixes

uklotzde opened the pull request:
The harvest from the tag writing branch with some lessons learned. Especially the handling of ID3 tags is always tricky. I already use this stuff heavily to keep the tags of my files (mostly MP3/M4A/FLAC) in sync with the Mixxx library, including bpm, key, and replay gain.

iTunes still doesn't recognize the ID3v2 comment tag in MP3 files written by TagLib. Give me a hint if you know what's missing ;)

uklotzde committed:
Track metadata: Replace QScopedPointer with std::unique_ptr

...and some minor changes for debugging purposes

uklotzde committed:
Track metadata: Add functions for normalizing floating-point values

uklotzde committed:
Track metadata: Add function for formatting duration values

uklotzde committed:
Track metadata: Read the first non-empty entry of a frame/string list

uklotzde committed:
Track metadata: Fix reading/writing of ID3v2 comments

daschuer commented on src/metadata/trackmetadata.h:
```diff
@@ -90,6 +90,14 @@ class TrackMetadata {
         m_key = key;
     }
 
+    // URL (used in promo track)
```
We have removed the promo trackside a while ago. So this comment is outdated.

daschuer commented on src/metadata/trackmetadatataglib.cpp:
```diff
@@ -148,22 +152,22 @@ inline QString toQString(const TagLib::ID3v2::Frame& frame) {
-    } else {
-        return toQString(*frameList.front());
+QString toQStringFirstNotEmpty(
+        const TagLib::ID3v2::FrameList& frameList) {
```
I think I can remember, that we had similar discussion before, but I cannot remember the details. 
(Who cares, we can discuss this new, with the new data on this new PR. :-P )

We have a string list, when there are more than one tags of a requested type in the file, right?

We should be sure about the reasons and why we skip a valid empty value. 

Does it happen when a file has two sets of meta-data? I can think of a situation when one application writes the first set and a second application writes the second. I this case, we may want to display the metadata from on application only.

What are your use cases? What are the realistic reasons for string-lists?

uklotzde replied:
StringLists are necessary for multi-valued fields that use multiple frames 
or items with the same id or key. Multiple values in a single frame are also 
supported if the standard defines a value separator like '\0' (e.g. ID3v2.4).

Some frame types additionally support a "description" to distinguish 
multiple subtypes. This is the reason why we need to read only from comment 
frames with an empty description! Otherwise the comment frames get mixed up 
and this is what TagLib does if you simply use Tag::set/getComment(). Not 
very smart, actually.

Reading the first non-empty frame is just a little bit smarter than only 
reading always the first frame of the list. It doesn't really matter, only 
for rare cases. I apply the same strategy when writing back any data, so we 
won't overwrite any data that has not been read before. Keeping the files in 
sync with the Mixxx library is required in both cases.

On 10/20/2015 08:47 PM, Daniel Schürmann wrote:
>
> In src/metadata/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/750#discussion_r42536999>:
>
> > @@ -148,22 +152,22 @@ inline QString toQString(const TagLib::ID3v2::Frame& frame) {
> >  }
> >
> >  // Returns the first frame of an ID3v2 tag as a string.
> > -inline QString toQStringFirst(const TagLib::ID3v2::FrameList& frameList) {
> > -    if (frameList.isEmpty() || (NULL == frameList.front())) {
> > -        return QString();
> > -    } else {
> > -        return toQString(*frameList.front());
> > +QString toQStringFirstNotEmpty(
> > +        const TagLib::ID3v2::FrameList& frameList) {
>
> I think I can remember, that we had similar discussion before, but I 
> cannot remember the details.
> (Who cares, we can discuss this new, with the new data on this new PR. :-P )
>
> We have a string list, when there are more than one tags of a requested 
> type in the file, right?
>
> We should be sure about the reasons and why we skip a valid empty value.
>
> Does it happen when a file has two sets of meta-data? I can think of a 
> situation when one application writes the first set and a second 
> application writes the second. I this case, we may want to display the 
> metadata from on application only.
>
> What are your use cases? What are the realistic reasons for string-lists?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/750/files#r42536999>.
>

uklotzde committed:
Track metadata: Reorder fields

...and some minor changes

uklotzde committed:
Track metadata: Add operators == and !=

daschuer commented:
Thank you for you explaining.

During test, I experience a (probably unrelated) dead lock in 
constrainBpm()
when bpm is not a value

daschuer commented:
Plain master does not suffer the issue. The effected track was an m4a track.

uklotzde commented:
Would be interesting to know if some special case handling is missing that 
can be verified with a unit test.

The unit tests for normalizeBpm()/normalizeReplayGain() were really helpful 
and I found such a corner case for 0 dB.

On 10/22/2015 11:48 AM, Daniel Schürmann wrote:
>
> Plain master does not suffer the issue. The effected track was an m4a track.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/750#issuecomment-150164381>.
>

uklotzde committed:
Track metadata: Encapsulate replay gain properties in their own class

uklotzde committed:
Track metadata: Reuse new ReplayGain class in TrackMetadata

uklotzde committed:
Track metadata: Handle replay gain 0 dB at file metadata level

The special case handling is related to file tags and should not be part
of the generic replay gain implementation.

daschuer commented:
LGTM. Thank you.
